### ISSUE TITLE
Clustering performance

### ISSUE THREAD

**Issue by krypt-n** [VROOM-Project/vroom#111]
I profiled the optimization of a fairly large CVP instance (http://vrp.atd-lab.inf.puc-rio.br/media/com_vrp/instances/X/X-n1001-k43.vrp) and noticed that the `std::make_heap` operations in `clustering.cpp` were responsible for a lot of the time taken to solve the instance. If I understand the usage of `std::make_heap` in the clustering algoritihms correctly, it is used to get the best possible candidate from the candidates vector. After that the heap is invalidated since the costs and regrets change. That's why the heap can't be reused between iterations and the repeated creation of the heap ends up  being pretty expensive.

Since the clustering algorithms only need the best element from the candidate array, `std::nth_element` can be used instead. `std::nth_element` avoids a lot of the work of `std::make_heap` and additionally is completely inlined by the compiler, resulting in the following improvement of runtime:

Before:
`"summary":{"cost":82940,"unassigned":0,"computing_times":{"loading":64,"solving":2785}}`

After:
`"summary":{"cost":82691,"unassigned":0,"computing_times":{"loading":65,"solving":1283}}`

I'm a bit worried by the difference in solution cost. It could just be an indicator of different stability properties of `std::nth_element` vs `std::make_heap`, but I'm not sure that I didn't miss anything. I'll try some more problem instances and open a PR.

I understand that the runtime of the clustering will be insignificant compared to the runtime of the local search phase, but this was a very low-hanging performance improvement.

**Comment by krypt-n** [comment]
The performance comparison with all instances from Uchoa et al. (2014) is quite impressive:
Before (all instances added together):
`{'distance': 0, 'computing_times': {'solving': 110620, 'loading': 1509}, 'cost': 7296615}`

After (all instances added together):
`{'distance': 0, 'computing_times': {'solving': 56525, 'loading': 1507}, 'cost': 7257162}`

However, the cost difference is quite visible here and definitely needs to be investigated.

**Comment by krypt-n** [comment]
Ah, it was indeed a sorting stability issue. If I fix the ordering manually with a modified `eval_lambda`, the results are the same:
```
    auto eval_lambda = [&](auto i, auto j) {
      double val1 = regret_coeff * static_cast<double>(regrets[v][i]) -
               static_cast<double>(costs[i]);
      double val2 = regret_coeff * static_cast<double>(regrets[v][j]) -
               static_cast<double>(costs[j]);
      if (val1 == val2) {
        return i < j;
      } else {
        return val1 < val2;
      }
    };
```

Without `std::nth_element`:
`{'distance': 0, 'computing_times': {'solving': 119802, 'loading': 1478}, 'cost': 7277582}`

With `std::nth_element`:
`{'distance': 0, 'computing_times': {'solving': 62511, 'loading': 1485}, 'cost': 7277582}`

Note that the whole thing got a bit slower since the modified `eval_lambda` introduces a new branch in a very hot part of the code.

Sorry for the walls of text in this issue. The question is: do you want a PR for the `std::nth_element` optimisation although it changes solution costs slightly?

**Comment by jcoupey** [comment]
Thanks for investigating this. Yes, you got it right: we only need to pick the best candidate at each step, like in the basic Prim algorithm for minimum spanning tree. The thing here is that, as you noticed, costs are updated throughout the procedure.

I didn't know `std::nth_element` but it sounds just like it could strip out the unnecessary work from `std::make_heap` in our case, so happy to have a PR with your changes.

One thing I didn't look into is how (or if) this would impact popping jobs from the heap. This happens on two situations:
1. a picked job has to be removed from the heap (or other clusters heaps for the parallel version)
2. we pop all candidates from all heaps when none is valid regarding capacities

**Comment by jcoupey** [comment]
Regarding the cost difference, it's fine if it's only a matter of sorting stability. We always make greedy choices anyway so there is no reason to favor stability, especially if it slows things down.

When you work at the clustering level, beware that the costs reported after solving can be puzzling because they are actually just the costs of the subsequent TSP routes. The clustering costs are the sum of the spanning trees edges so they have no actual route meaning, they just provide a bound for the final cost. AFAICT from my experiments, the smaller the clustering cost, the smaller the final cost (usually) is. But I've seen edge cases where a cheaper clustering leads to a higher cost after the TSP phase.
If you want to monitor the clustering costs, there is still a lot of trace stuff you can get in output using `-V`, then pipe to `grep clustering`.

**Comment by jcoupey** [comment]
When running single-threaded on all instances from Uchoa et al., I'm getting a total solving time of 216430 ms with the base branch (at da628ea), down to 92212 ms with #112. More than a 2x speedup! :+1: 
The speedup is less important when running multi-threaded but still significant. As mentioned above, changes in solution quality related to sorting stability are not to be worried about.

> I understand that the runtime of the clustering will be insignificant compared to the runtime of the local search phase, but this was a very low-hanging performance improvement.

Right, but this is still an important improvement because:

- as problem sizes grow, the clustering heuristic duration will become really significant at some point, so we're pushing the limit
-  the set of clustering parameters used [here](https://github.com/VROOM-Project/vroom/blob/master/src/problems/cvrp/cvrp.cpp#L34-L51) will have to be tuned. This speedup allows to either keep the same parameter coverage with faster solutions, or use more parameters with no extra computing cost.

**Comment by jcoupey** [comment]
Solved by #112.
